src/backend/executor/execMain.c

```diff
@@ -4603,10 +4604,27 @@ targetid_get_partition(Oid targetid, EState *estate, bool openIndices)
 		if (openIndices)
 			ExecOpenIndices(childInfo);
 
-		map_part_attrs(parentInfo->ri_RelationDesc,
-					   childInfo->ri_RelationDesc,
-					   &(childInfo->ri_partInsertMap),
-					   TRUE); /* throw on error, so result not needed */
+		/*
+		 * es_result_relations does not always represent the parent relation.
+		 * E.g. planner's UPDATE command on parent partition leads to multiple
+		 * subplans and result relations due to preceding inheritance planning.
+		 * In this case es_result_relations points to one of the partitions, not
+		 * to the parent. Thus, the descriptor mapping should be performed only
+		 * for the case if es_result_relations really corresponds to the parent.
+		 * Otherwise, there is a chance to reconstruct already valid tuple and
+		 * get the wrong results (e.g. target partition relation descriptor is
+		 * different from parentInfo's, but it's UPDATE (legacy planner) and
+		 * parentInfo represents another partition, which is not the true
+		 * parent). Moreover, if we are initially modify a leaf partition,
+		 * i.e we called a DML command straight on child partition, or it's
+		 * inheritance plan execution, the tuple descriptor already matches
+		 * the partition's, and the extra mapping is unnecessary.
+		 */
+		if (RelationGetRelid(parentInfo->ri_RelationDesc) == parentRelid)
+			map_part_attrs(parentInfo->ri_RelationDesc,
+						   childInfo->ri_RelationDesc,
+						   &(childInfo->ri_partInsertMap),
+						   TRUE); /* throw on error, so result not needed */
```

Comment:
this logic handles the insert into root partition. right? why don't we check if we need ri_partInsertMap or not? after all, if there are no deleted columns, then the map is unnecessary (right?). this check is done in makePartitionCheckMap.
https://github.com/arenadata/gpdb/blob/f6aa64215a15482e98fbd19b0dc73e79a78604af/src/backend/executor/nodeModifyTable.c#L2495

Reply:
Well, if there are no dropped columns, the `map_part_attrs` simply won't build the ` ri_partInsertMap`, because the `parentInfo->ri_RelationDesc` and  `childInfo->ri_RelationDesc` are equal.